compiler/compiler_test.go

```diff
@@ -484,19 +490,363 @@ func TestCompileAndEval(t *testing.T) {
+							{Key: &semantic.Identifier{Name: "r"}},
+						},
+					},
+					Body: &semantic.Block{
+						Body: []semantic.Statement{
+							&semantic.NativeVariableAssignment{
+								Identifier: &semantic.Identifier{Name: "v"},
+								Init: &semantic.ConditionalExpression{
+									Test: &semantic.BinaryExpression{
+										Operator: ast.LessThanOperator,
+										Left:     &semantic.IdentifierExpression{Name: "r"},
+										Right:    &semantic.IntegerLiteral{Value: 0},
+									},
+									Consequent: &semantic.UnaryExpression{
+										Operator: ast.SubtractionOperator,
+										Argument: &semantic.IdentifierExpression{Name: "r"},
+									},
+									Alternate: &semantic.IdentifierExpression{Name: "r"},
+								},
+							},
+							&semantic.ReturnStatement{
+								Argument: &semantic.BinaryExpression{
+									Operator: ast.MultiplicationOperator,
+									Left:     &semantic.IdentifierExpression{Name: "v"},
+									Right:    &semantic.IdentifierExpression{Name: "v"},
+								},
+							},
+						},
+					},
+				},
+			},
+			inType: semantic.NewObjectType(map[string]semantic.Type{
+				"r": semantic.Int,
+			}),
+			input: values.NewObjectWithValues(map[string]values.Value{
+				"r": values.NewInt(-3),
+			}),
+			want:    values.NewInt(9),
+			wantErr: false,
+		},
+		{
+			name: "array literal",
+			// f = () => [1.0, 2.0, 3.0]
+			fn: &semantic.FunctionExpression{
+				Block: &semantic.FunctionBlock{
+					Parameters: &semantic.FunctionParameters{},
+					Body: &semantic.ArrayExpression{
+						Elements: []semantic.Expression{
+							&semantic.FloatLiteral{Value: 1},
+							&semantic.FloatLiteral{Value: 2},
+							&semantic.FloatLiteral{Value: 3},
+						},
+					},
+				},
+			},
+			inType: semantic.NewObjectType(nil),
+			input:  values.NewObjectWithValues(nil),
+			want: values.NewArrayWithBacking(
+				semantic.Float,
+				[]values.Value{
+					values.NewFloat(1),
+					values.NewFloat(2),
+					values.NewFloat(3),
+				},
+			),
+			wantErr: false,
+		},
+		{
+			name: "logical expression",
+			// f = (a, b) => a or b
+			fn: &semantic.FunctionExpression{
+				Block: &semantic.FunctionBlock{
+					Parameters: &semantic.FunctionParameters{
+						List: []*semantic.FunctionParameter{
+							{Key: &semantic.Identifier{Name: "a"}},
+							{Key: &semantic.Identifier{Name: "b"}},
+						},
+					},
+					Body: &semantic.LogicalExpression{
+						Operator: ast.OrOperator,
+						Left:     &semantic.IdentifierExpression{Name: "a"},
+						Right:    &semantic.IdentifierExpression{Name: "b"},
+					},
+				},
+			},
+			inType: semantic.NewObjectType(map[string]semantic.Type{
+				"a": semantic.Bool,
+				"b": semantic.Bool,
+			}),
+			input: values.NewObjectWithValues(map[string]values.Value{
+				"a": values.NewBool(true),
+				"b": values.NewBool(false),
+			}),
+			want:    values.NewBool(true),
+			wantErr: false,
+		},
+		{
+			name: "call with nonexistant value",
+			// f = (r) => r.a + r.b
```

Comment:
I would not have expected this test case to type check because `r.b` has to type - it's null. If you replicate this test case in the semantic package (type inference test), would it pass there too?

Reply:
I'll try writing the tests there tomorrow, but the cool part is it doesn't actually have a type because the value isn't present in the record value. Since type checking happens before the `values.Value` stuff, null isn't technically present anywhere and the type checking seems to pass around the `Tvar` everywhere so weird things like this work.

```
r = {a: 1}
a = r.b
c = a + 2
```

This ends up correctly mapping `a` to be an integer because `+` sets the type constraint and there's nothing else to contradict it.

The problem seems to be when an operator doesn't set a type constraint like the comparison operators. They fail because they can't infer the type from something that doesn't exist so we'll have to find some way to resolve `Tvar`'s with a default value possibly. I think constraints may work for that, but I need to do more investigation.

Reply:
Oh that's very interesting. I did not think about that. So basically the worst that can happen is that a type variable just stays around and never gets assigned to a concrete type. But there's still no contradiction or anything, so it passes type checking. That's really cool.

I don't think it should work any differently for comparison operators. In fact, I just confirmed that those operators don't constrain their operands to be of equal types, which is definitely a bug. You should not be able to compare values of different types.

Reply:
Here's the issue describing the type checking bug you found for comparison operators #1356. Feel free to close it out with these changes.